slow_odgi/examples.md

```diff
@@ -0,0 +1,157 @@
+# Pollen language walkthrough
+
+## Core types in Pollen
+
+    type Graph = {
+      segments: Set<Segment>;
+      paths: Set<Path>;
+      links: Set<Links>; 
+    }
+
+    type Segment = {
+      sequence: Strand; //ACTGAC, etc. 
+      links: Set<Link>; //links encode their direction + orientation
+      steps: Set<Step>; //steps that go through segment
+    }
+
+    type Handle = {
+      segment: Segment; 
+      orientation: bool;
+    }
+
+    type Path = {
+      steps: List<Step>;
+    }
+
+    type Step = {
+      path: Path; 
+      idx: int; //where in the path is this sequence?
+      handle: Handle; //segment + orientation
+    }
+
+    type Base = A | C | T | G | N
+    type Strand = list Base
+
+    type Link = {
+      start: Handle; 
+      end: Handle;
+    }
+
+## Chop
+
+## Crush
+
+    graph out_g;
+    parset out_segs[Segment, out_g];
+
+    for segment in Segments {
+      seq = Strand();
+      in_n = false;
+      for c in segment.sequence {
+        if c == 'N' {
+          if !in_n {
+            in_n = true;
+            seq.push(c);
+          }
+        } else {
+            in_n = false;
+            seq.push(c);
+        }
+      }
+      emit { segment with sequence: seq } to out_segs;
+    }
+
+## Degree
+
+    // Output is a (Segment, int) array.
+    outset out_degs[Segment, int];
+
+    for segment in in_g.segments {
+      emit (segment, segment.edges.size()) to out_degs;
+    }
+
+## Depth
+
+    // Output is a (Segment, int) array.
+    outset out_depths[Segment, int];
+
+    for segment in in_g.segments {
+      emit (segment, segment.steps.size()) to out_depths;
+    }
+
+## Flatten
+
+## Flip
+
+    graph out_g;
+    parset out_steps[Step, out_g];
+    parset out_paths[Path, out_g];
+
+    local is_rev(Path path) {
+      fw = 0;
+      bw = 0;
+      for step.handle as sh in path.steps {
```

Comment:
I'm a little confused about this `for x as y in z` syntax. What is `x` allowed to be—for example, in this situation, what does `step` refer to? I think this stuff may be equivalent to:

```
for step in path.steps {
  sh = step.handle;
  ...
}
```

Is that right? If so, I would love to stick to that simpler use of `for`. I think putting a "dot expression" like `step.handle` in the "binding" position of the loop raises a lot of questions about how we want the program to behave.

Reply:
Noted, definitely took streamlining the code a bit far there and obfuscated things. Will revert to your suggestion above!